Add blog post "Accessing Apisix-dashboard from everywhere with Keycloak authentication" (#1468)
iresty/iresty.github.io: blog/en/blog/2023/01/02/accessing_apisix-dashboard_from_everywhere_with_keycloak_authentication.md

---

> title: "Accessing Apisix-dashboard from everywhere with Keycloak authentication"

Reviewer: Hi, please use `APISIX` instead of `Apisix` and check all other places. :) We only use `Apisix` in Ingress Resources like `ApisixRoute`.

Author: @juzhiyuan  Please help: this is the first time I do a revision in github
Where have I to do the changes? 
Here: https://github.com/MirtoBusico/apisix-website/blob/master/blog/en/blog/2023/01/02/accessing_apisix-dashboard_from_everywhere_with_keycloak_authentication.md

And after the modifications simply commit?

Author: I suppose I have catched alla APISIX occurrencies :)

> title: "Accessing Apisix-dashboard from everywhere with Keycloak authentication"

Reviewer 2: ```suggestion
title: "Accessing APISIX-Dashboard from Everywhere with Keycloak Authentication"
```

authors:
  - name: "Busico Mirto Silvio"
    title: ""
    url: "https://github.com/MirtoBusico"
    image_url: "https://avatars.githubusercontent.com/u/11090934?s=400&u=644e4f87c2fad56760f6eb4f46cbcb4db059880a&v=4"
keywords:
  - Apache APISIX
  - API Gateway
  - Authentication
  - OpenID Connect
  - Keycloak
  - Nginx reverse proxy
  - openid-connect
description: This article describes how to expose Apisix-dashboard using Apisix to authenticate the access using openid-connect plugin and a Keycloak server to manage identities
tags: [Authentication, Plugins]

> image: https://github.com/MirtoBusico/assets-for-blogs/blob/main/blog01a.png

Reviewer 2: ```suggestion
keywords:
  - Alternatives to NGINX
  - Access and Authentication
  - Nginx reverse proxy
  - OpenID-Connect
  - Keycloak
description: This guest blog shares how to expose the APISIX Dashboard using APISIX to authenticate access with the OpenID-Connect plugin and Keycloak server to manage identities.
tags:
  - APISIX Basics
  - API Gateway Concepts
  - Authorization & Authentication
cover: https://github.com/MirtoBusico/assets-for-blogs/blob/main/blog01a.png
```

---

> This article describes how to setup an external access to apisix-dashboard protecting the URL with authentication managed by a keycloak server.

<!--truncate-->

# Accessing Apisix-dashboard from everywhere with Keycloak authentication

> 

Reviewer 2: ```suggestion
```

![framework](https://github.com/MirtoBusico/assets-for-blogs/blob/main/blog01a.png)

This article presents how to setup a framework where a user can access the Apisix-dashboard protected using an authentication system managed by a Keycloak server.

> # Prerequisites

Reviewer 2: ```suggestion
## Prerequisites
```

|   hkm       | Kubernetes master     | master node         | control plane manager for K8S            |
|   hkw1      | K8S worker 1    | first worker node         |     node for hosting pods        |
|   hkw2      | K8S worker 2    | second worker node         |     node for hosting pods        |
|   hkw3      | K8S worker 3    | third worker node         |     node for hosting pods        |

The **hserv** VM have two lan cards: one on an external lan to expose services and one an internal lan to communicate with the Kubernetes (from now K8S) cluster.
All the other VM are only connected to the internal lan.

All the machines resolve the IP addresses using the DNS server installed on **hserv**

**Hserv** and **hdev** machines have a Graphical User Interface (from now GUI). All the other machines have only the character console.

> The real framework is more complex. Here are reported only the relevant components

All machines use Ubuntu distribution but commands reported here should worh for other distributions with some modifications.
The username used throughout this article will be **"sysop"** So the home directory will be indicated as **"/home/sysop"** or **"~/"**.

> # Create a Certification authority and certificates

Reviewer 2: ```suggestion
## Create a Certification Authority and Certificates
```

For all the VM the DNS server will resolve **"apisix.h.net"** to the external address of **hserv**.
In all others machine that will access the the services exposed by **hserv** there will be a line in the **"/etc/hosts"** file resolving **"apisix.h.net"** to the external address of **hserv**.

> Working on **hserv**

Create the directory for the entire project software
```
cd
mkdir H
```
Create the directory to hold the Certification authority (from now CA) certificates and the web sites certificates
```
cd ~/H
mkdir hservcerts
cd hservcerts
```
Create a private key for **"hservca"**
```
sudo openssl genrsa -out hservca.key 2048
```
This generates a **hservca.key** key file. Using this fiile generate the CA certificate
```
sudo openssl req -x509 -new -nodes -key hservca.key -sha256 -days 3650 -out hservca.pem
```
This generates a **hservca.pem"** certificate file. These two files will be used to create the web sites certificates

> ## Add the CA to browsers

Reviewer 2: ```suggestion
### Add the CA to Browsers
```

To be able to access the web sites with certidicates issued by this private CA, the CA certificate file have to be added to the web browser that will access these sites.

> Working on hdev

Copy the **"hservca.pem"** file in any machine that will access these sites.
```
cd
cp ~/H/hservcerts/hservca.pem .
rcp hservca.pem mirto@_any_machine_name_://home/_your_username_/
```

For **Firefox** browser go to:
```
Preferences -> Privacy & Security -> Certificates -> View Certificates -> Authorities -> Import 
```
and import **"hservca.pem"** (remember to flag all options)

For **Chromium** or **Chrome** browsers go to:
```
Settings -> Advanced -> Privacy and security -> Manage certificates -> Authorities -> Import (flag all options)
```
and import **"hservca.pem"** (remember to flag all options)

> ## Add the CA to the operating system

Reviewer 2: ```suggestion
### Add the CA to the Operating System
```

> Working on hdev
 
Copy the "hservca.pem" file in the "/home/sysop" directory.
Copy this file on any other machine that will use sertificates signed by this CA.
```
cd
cp ~/H/hservcerts/hservca.pem .
rcp hservca.pem mirto@_any_machine_name_://home/_your_username_/
```

> Work on any machine

Then on any machine and hserv do the following:

```
cd
sudo mkdir -p /usr/share/ca-certificates/extra
sudo cp hservca.pem /usr/share/ca-certificates/extra/hservca.crt
sudo dpkg-reconfigure ca-certificates
```
> **Attention**:
> 
>    • "dpkg-reconfigure ca-certificates" do not recognize the ".pem" extension. Copy the **"hservca.pem"** file to **"hservca.crt"**
>    
>    • select the new certificate in "dpkg-reconfigure ca-certificates" (extra/hservca.crt is not selected)

Confirm that you want to proceed: select “yes” and click “Ok”. Select the new “hservca.crt” entry and click “Ok”

![confirm](https://github.com/MirtoBusico/assets-for-blogs/blob/main/ca-certificates.png)

> # Install nginx-mainline

Reviewer 2: ```suggestion
## Install nginx-mainline
```

Verify the system is updated
```
sudo apt update
sudo apt full-upgrade
```
Install prerequisites
```
sudo apt install wget gnupg2 ca-certificates lsb-release ubuntu-keyring software-properties-common -y
```
Download the Nginx GPG key
```
wget -O- https://nginx.org/keys/nginx_signing.key | sudo gpg --dearmor | sudo tee /usr/share/keyrings/nginx-archive-keyring.gpg
```
Add the Nginx mainline apt repository
```
echo deb [arch=amd64,arm64 signed-by=/usr/share/keyrings/nginx-archive-keyring.gpg] http://nginx.org/packages/mainline/ubuntu `lsb_release -cs` nginx | sudo tee /etc/apt/sources.list.d/nginx-mainline.list
```
Pin the Nginx repository
```
echo -e "Package: *\nPin: origin nginx.org\nPin: release o=nginx\nPin-Priority: 900\n" | sudo tee /etc/apt/preferences.d/99nginx
```
Update apt and install nginx
```
sudo apt update
sudo apt install nginx
```

> # Install Keycloak
> > Work on **hserv**
> 
> ## Prerequisites
> Install jdk
> ```
> sudo apt install default-jdk
> ```
> Remove anacron
> ```
> sudo apt remove anacron
> ```
> Reboot the **hserv** machine
> 
> ## Keycloak installation
> Go in base installation directory and get keycloak installation files (verify what is the last release)

Reviewer 2: ```suggestion
## Install Keycloak
> Work on **hserv**

### Prerequisites
Install jdk
```
sudo apt install default-jdk
```
Remove anacron
```
sudo apt remove anacron
```
Reboot the **hserv** machine

### Keycloak Installation
Go in base installation directory and get keycloak installation files (verify what is the last release)
```

```
cd ~/H/
wget https://github.com/keycloak/keycloak/releases/download/20.0.1/keycloak-20.0.1.zip
```
Extract the files
```
unzip keycloak-20.0.1.zip
```
Go to the bin directory and start keycloak in standalone mode
```
cd ~/H/keycloak-20.0.1/bin/
./kc.sh start-dev
```
Verify that Keycloak is accessible from **hserv** at the URL **"http://localhost:8080"**

Create the admin user as name **"admin"** and password **"1357Togo"**

![k6k01](https://github.com/MirtoBusico/assets-for-blogs/blob/main/k6k01.png)

Go to the administration console

![k6k02](https://github.com/MirtoBusico/assets-for-blogs/blob/main/k6k02.png)

Login and the “Master” realm appears. Note the Keycloak version

![k6k03](https://github.com/MirtoBusico/assets-for-blogs/blob/main/k6k03.png)

> ## Automatic Keycloak startup

Reviewer 2: ```suggestion
### Automatic Keycloak Startup
```

> Work on **hserv**

Create in **“/usr/lib/systemd/system”** a file named **“keycloak.service”** containing
```
[Unit]
Description=keycloak service
After=network.service

[Service]
ExecStart=/home/sysop/H/keycloak-20.0.1/bin/kc.sh start-dev >/var/log/keycloak.log 2>&1
PIDFile=/var/run/keycloak.pid

[Install]
WantedBy=multi-user.target
```
Enable and activate the service
```
sudo systemctl enable keycloak
sudo systemctl start keycloak
```
Reboot hserv and verify Keycloak is accessible at startup

> ## Create site and certificates for "https://k6k.h.net"

Reviewer 2: ```suggestion
### Create Site and Certificates for "https://k6k.h.net"
```

```
sudo mkdir /usr/share/nginx/k6k
sudo chmod 777 /usr/share/nginx/k6k
vi /usr/share/nginx/k6k/index.html
```
Put in index.html filke the "K6K" base document
```
<!DOCTYPE html>
<html>
<text>

<h1>K6K default page</h1>

</text>
</html>
```
In the directory **“/etc/nginx/conf.d”** create the file **“k6k.conf”**
```
cd /etc/nginx/conf.d
sudo vi k6k.conf
```
The file contains
```
server {

    listen 443 ssl http2;
    server_name k6k.h.net;
    root   /usr/share/nginx/k6k;
    
    access_log  /var/log/nginx/k6k.access.log;
    error_log  /var/log/nginx/k6k.error.log;

    ssl_certificate     /home/sysop/H/hservcerts/k6k.crt;
    ssl_certificate_key /home/sysop/H/hservcerts/k6k.key;

    
    location / {
        index  index.html index.htm;
    }

    error_page  404              /404.html;

    # redirect server error pages to the static page /50x.html
    #
    error_page   500 502 503 504  /50x.html;
    location = /50x.html {
        root   /usr/share/nginx/html;
    }

}
```
Restart Nginx
```
sudo systemctl restart nginx
```
Try to access “https://k6k.h.net” from a browser: the k6k base document will be showed

> ## Add keycloak reverse proxy

Reviewer 2: ```suggestion
### Add Keycloak Reverse Proxy
```

    proxy_set_header Host $host;
    proxy_set_header X-Real-IP $remote_addr;
    proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
    proxy_set_header X-Forwarded-Host $host;
    proxy_set_header X-Forwarded-Proto $scheme;
    
    location / {
        proxy_pass http://127.0.0.1:8080;
    }

    error_page  404              /404.html;

    # redirect server error pages to the static page /50x.html
    #
    error_page   500 502 503 504  /50x.html;
    location = /50x.html {
        root   /usr/share/nginx/html;
    }

}
```
Restart Nginx
```
sudo systemctl restart nginx
```
Rebuild keycloak for production
```
cd ~/H/keycloak-20.0.1/bin/
./kc.sh --verbose build
```

Change in **“/usr/lib/systemd/system”** the file named **“keycloak.service”** with this content
```
[Unit]
Description=keycloak service
After=network.service

[Service]
ExecStart=/home/sysop/H/keycloak-20.0.1/bin/kc.sh start --proxy edge --hostname-strict=false >/var/log/keycloak.log 2>&1
PIDFile=/var/run/keycloak.pid

[Install]
WantedBy=multi-user.target
```
Enable and activate the service
```
sudo systemctl daemon-reload
sudo systemctl restart keycloak
```

> # Apisix api gateway
> 
> ## Addresses for apisix-dashboard

Reviewer 2: ```suggestion
## APISIX API Gateway

### Addresses for APISIX Dashboard
```

The address used is the exsternal address o the **hsrv** machine

In the **"/etc/hosts"** file of any machine accessing the cluster through the nginx reverse proxy add the line
```
192.168.100.20 apisix apisix.h.net
```

In the DNS server on **hserv** add the apisix entry in the **“h.net”** DNS zone with address **“192.168.100.20”**

![dns02](https://github.com/MirtoBusico/assets-for-blogs/blob/main/dns02.png)

> ## Apisix deployment

Reviewer 2: ```suggestion
### APISIX Deployment
```

- enable (set to true) apisix dashboard and ingress-controller

The relevant tiles changed are:

```
gateway:
  type: LoadBalancer
...
  tls:
    enabled: true
    servicePort: 443
    containerPort: 9443
    existingCASecret: "hservcacert"
    certCAFilename: "cert"
...
discovery:
  enabled: true
  registry:
    dns:
        servers:
            - "10.43.0.10:53"
...
   httpSrv: |
    set $session_secret 0123456789a5bac9bb3c868ec8b202e93;
...

dashboard:
  enabled: true

ingress-controller:
  enabled: true
```

Install apisix using the new values.yaml file
```
helm install apisix apisix/apisix -f apisix-values.yaml \
--set ingress-controller.config.apisix.serviceNamespace=apisix \
--set ingress-controller.config.apisix.serviceName=apisix-admin \
--set ingress-controller.config.kubernetes.apisixRouteVersion=apisix.apache.org/v2beta3 \
--namespace apisix
```

Wait for the pods to start (it can take some time)
```
kubectl -n apisix wait --for=condition=Ready pods --all
kubectl get pods -n apisix
```

When all the Apisix pods will be in **Running** state the installation is completed

> ##  Accessing apisix dashboard

Reviewer 2: ```suggestion
###  Accessing APISIX Dashboard
```

> Work on **hdev**

Port forward apisix-dashboard
```
kubectl -n apisix port-forward service/apisix-dashboard 9090:80
```
The command output should be something like
```
sysop@hdev:~$ kubectl -n apisix port-forward service/apisix-dashboard 9090:80
Forwarding from 127.0.0.1:8080 -> 9000
Forwarding from [::1]:8080 -> 9000
```

Then access the dashboard on **“hdev”** pointing the web browser to the url “http://localhost:9090”
Login with **“admin / admin”**

![ad01](https://github.com/MirtoBusico/assets-for-blogs/blob/main/ad01.png)

Verify the dashboard version

![ad02](https://github.com/MirtoBusico/assets-for-blogs/blob/main/ad02.png)

> ## Create Apisix resources for apisix-dashboard
> 
> ### Create the certificate for **"apisix.h.net"**

Reviewer 2: ```suggestion
### Create APISIX Resources for APISIX Dashboard

#### Create the Certificate for **"apisix.h.net"**
```

> Work on **hserv**

In the “~/H/hservcerts/” folder create the key and certificate for apisix.h.net

Create a file called **"apisixssl.cnf"** containing
```
[req]
default_bits = 2048
distinguished_name = req_distinguished_name
prompt = no

[req_distinguished_name]
C = IT
ST = Italy
L = Rome
O = Busico Mirto
OU = Laboratory
CN = apisix.h.net

[v3_ca]
subjectAltName = @alt_names

[alt_names]
DNS.1 = apisix
# other names
DNS.2 = apisix.h.net
DNS.3 = apisix.ext.h.net
DNS.4 = apisix.int.h.net
```
Create the server private key and csr
```
sudo openssl req -new -sha256 -nodes -newkey rsa:2048 -keyout apisix.key -out apisix.csr -config apisixssl.cnf
```

Create the certificate file.
```
sudo openssl x509 -req -in apisix.csr -CA hservca.pem -CAkey hservca.key -CAcreateserial -out apisix.crt -sha256 -days 3650 -extfile apisixssl.cnf -extensions v3_ca
```
Change the access rights for apisix key to permit nginx access
```
sudo chmod a+r apisix.key
```

> ### Apply certificates Nginx and enable HTTPS

Reviewer 2: ```suggestion
#### Apply Certificates NGINX and Enable HTTPS
```

```
cd /etc/nginx/conf.d
sudo vi apisix.conf
```
Put in the **“apisix.conf”** file this content
```
server {

    listen 443 ssl http2;
    server_name apisix.h.net;
    root   /usr/share/nginx/apisix;
    
    access_log  /var/log/nginx/apisix.access.log;
    error_log  /var/log/nginx/apisix.error.log;

    ssl_certificate     /home/sysop/H/hservcerts/apisix.crt;
    ssl_certificate_key /home/sysop/H/hservcerts/apisix.key;

    
    location / {
        index  index.html index.htm;
    }

    error_page  404              /404.html;

    # redirect server error pages to the static page /50x.html
    #
    error_page   500 502 503 504  /50x.html;
    location = /50x.html {
        root   /usr/share/nginx/html;
    }

}
```
Restart Nginx
```
sudo systemctl restart nginx
```
Add the apisix line in “/etc/hosts” on any machine that will access apisix-dashboard
```
192.168.100.20 apisix.h.net
```
Add the apisix A record in the DNS in “h.net” zone

![ad03](https://github.com/MirtoBusico/assets-for-blogs/blob/main/ad03.png)

Access “https://apisix.h.net” from a browser and the apisix default page should be presented

> ### Create the load balancer definition

Reviewer 2: ```suggestion
#### Create the Load Balancer Definition
```

    proxy_set_header X-Forwarded-Host $host;
    proxy_set_header X-Forwarded-Proto $scheme;

    location / {
        proxy_pass https://hcluster;
    }

    error_page  404              /404.html;

    # redirect server error pages to the static page /50x.html
    #
    error_page   500 502 503 504  /50x.html;
    location = /50x.html {
        root   /usr/share/nginx/html;
    }

}
```

> **Note:**
> 
> the lines
```
upstream hcluster {
    ip_hash;
    server 192.168.101.22:443;
    server 192.168.101.23:443;
    server 192.168.101.24:443;
}
```   
> use the internal address of the three kubernetes worker nodes and it is necessary to specify the 443 port to enable https traffic
>
> The lines
```
    proxy_busy_buffers_size   512k;
    proxy_buffers   4 512k;
    proxy_buffer_size   256k;
```
> are required because, after the Keycloak authentication, the apisix server replyes with the state in the URL.
> 
> With the default values nginx replies with a "response too big" error

Restart Nginx
```
sudo systemctl restart nginx
```

Access “https://apisix.h.net” from a browser. You should receive page not found error because there is no route in Apisix

> ### Create “apisix-dashboard” route and upstream with apisix-dashboard

Reviewer 2: ```suggestion
#### Create “apisix-dashboard” Route and Upstream with APISIX Dashboard
```

![ad05](https://github.com/MirtoBusico/assets-for-blogs/blob/main/ad05.png)

Then click “Next” and after click “Submit”

Click “View” to see the json upstream definition

![ad06](https://github.com/MirtoBusico/assets-for-blogs/blob/main/ad06.png)

Note the upstream name to be used in the next route definition

![ad07](https://github.com/MirtoBusico/assets-for-blogs/blob/main/ad07.png)

Now click “Create” on the “Route” page

![ad08](https://github.com/MirtoBusico/assets-for-blogs/blob/main/ad08.png)

Create a route ("Define api request" - on top): set name to “apisix-dashboard” and add a description

![ad09](https://github.com/MirtoBusico/assets-for-blogs/blob/main/ad09.png)

Create a route ("Define api request" - below): set host to **“apisix.h.net”** and path to **“/*”**. Then click **“Next”**

![ad10](https://github.com/MirtoBusico/assets-for-blogs/blob/main/ad10.png)

Select the previous defined “apisix” upstream from the dropdown list. Then click “Next”

![ad11](https://github.com/MirtoBusico/assets-for-blogs/blob/main/ad11.png)

For now don’t use plugins and click “Next”. Then click “Submit”

![ad12](https://github.com/MirtoBusico/assets-for-blogs/blob/main/ad12.png)

> ### Enable https in apisix

Reviewer 2: ```suggestion
### Enable HTTPS in APISIX
```

```
kubectl -n apisix port-forward service/apisix-dashboard 9090:80
```

Select the “SSL” page and click Create

![ad14](https://github.com/MirtoBusico/assets-for-blogs/blob/main/ad14.png)

Select **“Way: Upload”**, then click **“upload certificate”** and **“upload key”**. Clik “Next” (Take certificate and key files from **~/H/hservcerts**)

![ad15](https://github.com/MirtoBusico/assets-for-blogs/blob/main/ad15.png)

Preview the SSL resource and click “Submit”

![ad16](https://github.com/MirtoBusico/assets-for-blogs/blob/main/ad16.png)

The ssl resource appear in the list (note the SNI values)

![ad17](https://github.com/MirtoBusico/assets-for-blogs/blob/main/ad17.png)

Configure the “apisix-dashboard” route to enable http to https redirection

![ad18](https://github.com/MirtoBusico/assets-for-blogs/blob/main/ad18.png)

Set the **“Redirect”** field to **“Enable HTTPS”**. Then click “Next” until you see “Submit”. Click “Submit”

![ad19](https://github.com/MirtoBusico/assets-for-blogs/blob/main/ad19.png)

View the route configuration and verify that the redirect plugin is enabled

![ad20](https://github.com/MirtoBusico/assets-for-blogs/blob/main/ad20.png)

> Now from any machine you can access the apisix-dashboard at the url **“https://apisix.h.net”** nd verify that the apisix-dashboard login page is showed

Reviewer: ```suggestion
Now from any machine you can access the apisix-dashboard at the url **“https://apisix.h.net”** and verify that the apisix-dashboard login page is showed
```

![ad21](https://github.com/MirtoBusico/assets-for-blogs/blob/main/ad21.png)

Login with “admin” / “admin”

![ad22](https://github.com/MirtoBusico/assets-for-blogs/blob/main/ad22.png)

> ### Create keycloak definitions for openid-connect

Reviewer 2: ```suggestion
#### Create Keycloak Definitions for OpenID-Connect
```

![k6k06](https://github.com/MirtoBusico/assets-for-blogs/blob/main/k6k06.png)

Verify that the client protocol is **“openid-connect”** and click “Next”

![k6k07](https://github.com/MirtoBusico/assets-for-blogs/blob/main/k6k07.png)

Set “Client authentication” to "on" (means OIDC type confidential). Click “Save”

![k6k08](https://github.com/MirtoBusico/assets-for-blogs/blob/main/k6k08.png)

In **“Cient details”**, **“Settings”** tab, **“access settings”** section, set **“Valid redirect URI”** to **“*”**. Click **“Save”**

![k6k09](https://github.com/MirtoBusico/assets-for-blogs/blob/main/k6k09.png)

Create a new user 

![k6k10](https://github.com/MirtoBusico/assets-for-blogs/blob/main/k6k10.png)

Set the username to **“hcadmin”** and click **“Create”**

![k6k11](https://github.com/MirtoBusico/assets-for-blogs/blob/main/k6k11.png)

In the **“Credentials”** tab click **“Set password”** (same procedure for "Reset" password)

![k6k12](https://github.com/MirtoBusico/assets-for-blogs/blob/main/k6k12.png)

Set a permanent (Temporary set to Off) password to **“hcadmin”** (equal to the username)

![k6k13](https://github.com/MirtoBusico/assets-for-blogs/blob/main/k6k13.png)

> #### Get client ID and secret

Reviewer 2: ```suggestion
#### Get Client ID and Secret
```

Select **hcadmins** client. Go to **Credentials** tab; show the Secret and copy the client id and secret to be used in the next steps

client ID: **hcadmins**

Secret: **MoqLUhwgsEDi36II0KuJldKq4YGLHxl3**

![k6k14](https://github.com/MirtoBusico/assets-for-blogs/blob/main/k6k14.png)

In the **“Realm settings”**, **“General”** tab click on the link **“OpenID Endpoint Configuration”**

![k6k15](https://github.com/MirtoBusico/assets-for-blogs/blob/main/k6k15.png)

This is the shown page

![k6k16](https://github.com/MirtoBusico/assets-for-blogs/blob/main/k6k16.png)

Copy the link 
```
https://k6k.h.net/realms/hcluster_admins/.well-known/openid-configuration
```
Prepare a json client definition using the previous copied information
```
{
    "client_id":"hcadmins",
    "client_secret":"MoqLUhwgsEDi36II0KuJldKq4YGLHxl3",
    "discovery":"https://k6k.h.net/realms/hcluster_admins/.well-known/openid-configuration",
    "scope":"openid profile",
    "bearer_only":false,
    "realm":"hcluster_admins",
    "introspection_endpoint_auth_method":"client_secret_post",
    "redirect_uri":"https://apisix.h.net/*",
    "access_token_in_authorization_header":true
}
```

> ### Openid-connect for apisix dashboard

Reviewer 2: ```suggestion
#### OpenID-Connect for APISIX Dashboard
```

> Work on any machine

Configure the apisix-dashboard route.

Go to **“3 Plugin config”**

![ad23](https://github.com/MirtoBusico/assets-for-blogs/blob/main/ad23.png)

Click **“Enable”** on openid-connect plugin (if you have already defined the plugin you'll see "Edit" instead of "Enable")

![ad24](https://github.com/MirtoBusico/assets-for-blogs/blob/main/ad24.png)

Enable the plugin. Copy the previous defined json definition and click **“Submit”**

![ad25](https://github.com/MirtoBusico/assets-for-blogs/blob/main/ad25.png)

Clik "next" and then clik "Submit" to complete the route configuration

Then "view" the route to see the plugin configuration

![ad26](https://github.com/MirtoBusico/assets-for-blogs/blob/main/ad26.png)

> ## Accessing the protected apisix-dashboard route

Reviewer 2: ```suggestion
### Accessing the Protected APISIX Dashboard Route
```

> Work on any machine

Go to the URL
```
https://apisix.h.net
```
You will be redirected to the Keycloak login page for the "HCLUSTER_ADMINS" realm.
Login with the previous defined user "hcadmin" / "hcadmin"

![ad27](https://github.com/MirtoBusico/assets-for-blogs/blob/main/ad27.png)

The apisix-dashboard login will be presented. Login with "admin" / "admin"

![ad28](https://github.com/MirtoBusico/assets-for-blogs/blob/main/ad28.png)

And now you can see the apisix dashboard

![ad29](https://github.com/MirtoBusico/assets-for-blogs/blob/main/ad29.png)

> # Recap

Reviewer 2: ```suggestion
## Recap
```

---
title: "Accessing APISIX-Dashboard from Everywhere with Keycloak Authentication"
authors:
  - name: "Busico Mirto Silvio"
    title: ""
    url: "https://github.com/MirtoBusico"
    image_url: "https://avatars.githubusercontent.com/u/11090934?s=400&u=644e4f87c2fad56760f6eb4f46cbcb4db059880a&v=4"
keywords:
  - Alternatives to NGINX
  - Access and Authentication
  - Nginx reverse proxy
  - OpenID-Connect
  - Keycloak
description: This guest blog shares how to expose the APISIX Dashboard using APISIX to authenticate access with the OpenID-Connect plugin and Keycloak server to manage identities.
tags:
  - APISIX Basics
  - API Gateway Concepts
  - Authorization & Authentication

> cover: https://github.com/MirtoBusico/assets-for-blogs/blob/main/blog01a.png

Reviewer: ```suggestion
cover: https://static.apiseven.com/uploads/2023/01/19/FKDU7U6j_blog01a.png
```

|   hkm       | Kubernetes master     | master node         | control plane manager for K8S            |
|   hkw1      | K8S worker 1    | first worker node         |     node for hosting pods        |
|   hkw2      | K8S worker 2    | second worker node         |     node for hosting pods        |
|   hkw3      | K8S worker 3    | third worker node         |     node for hosting pods        |

The **hserv** VM have two lan cards: one on an external lan to expose services and one an internal lan to communicate with the Kubernetes (from now K8S) cluster.
All the other VM are only connected to the internal lan.

All the machines resolve the IP addresses using the DNS server installed on **hserv**

**Hserv** and **hdev** machines have a Graphical User Interface (from now GUI). All the other machines have only the character console.

> The real framework is more complex. Here are reported only the relevant components

All machines use Ubuntu distribution but commands reported here should worh for other distributions with some modifications.
The username used throughout this article will be **"sysop"** So the home directory will be indicated as **"/home/sysop"** or **"~/"**.

> ## Create a Certification authority and certificates

Reviewer 2: ```suggestion
## Create a Certification Authority and Certificates
```

For all the VM the DNS server will resolve **"apisix.h.net"** to the external address of **hserv**.
In all others machine that will access the the services exposed by **hserv** there will be a line in the **"/etc/hosts"** file resolving **"apisix.h.net"** to the external address of **hserv**.

> Working on **hserv**

Create the directory for the entire project software

```
cd
mkdir H
```

Create the directory to hold the Certification authority (from now CA) certificates and the web sites certificates

```
cd ~/H
mkdir hservcerts
cd hservcerts
```

Create a private key for **"hservca"**

```
sudo openssl genrsa -out hservca.key 2048
```

This generates a **hservca.key** key file. Using this fiile generate the CA certificate

```
sudo openssl req -x509 -new -nodes -key hservca.key -sha256 -days 3650 -out hservca.pem
```

This generates a **hservca.pem"** certificate file. These two files will be used to create the web sites certificates

> ### Add the CA to browsers

Reviewer 2: ```suggestion
### Add the CA to Browsers
```

To be able to access the web sites with certidicates issued by this private CA, the CA certificate file have to be added to the web browser that will access these sites.

> Working on hdev

Copy the **"hservca.pem"** file in any machine that will access these sites.

```
cd
cp ~/H/hservcerts/hservca.pem .
rcp hservca.pem mirto@_any_machine_name_://home/_your_username_/
```

For **Firefox** browser go to:

```
Preferences -> Privacy & Security -> Certificates -> View Certificates -> Authorities -> Import
```

and import **"hservca.pem"** (remember to flag all options)

For **Chromium** or **Chrome** browsers go to:

```
Settings -> Advanced -> Privacy and security -> Manage certificates -> Authorities -> Import (flag all options)
```

and import **"hservca.pem"** (remember to flag all options)

> ### Add the CA to the operating system

Reviewer 2: ```suggestion
### Add the CA to the Operating System
```

![confirm](https://github.com/MirtoBusico/assets-for-blogs/blob/main/ca-certificates.png)

## Install nginx-mainline

Verify the system is updated

```
sudo apt update
sudo apt full-upgrade
```

Install prerequisites

```
sudo apt install wget gnupg2 ca-certificates lsb-release ubuntu-keyring software-properties-common -y
```

Download the Nginx GPG key

```
wget -O- https://nginx.org/keys/nginx_signing.key | sudo gpg --dearmor | sudo tee /usr/share/keyrings/nginx-archive-keyring.gpg
```

Add the Nginx mainline apt repository

```
echo deb [arch=amd64,arm64 signed-by=/usr/share/keyrings/nginx-archive-keyring.gpg] http://nginx.org/packages/mainline/ubuntu `lsb_release -cs` nginx | sudo tee /etc/apt/sources.list.d/nginx-mainline.list
```

Pin the Nginx repository

```
echo -e "Package: *\nPin: origin nginx.org\nPin: release o=nginx\nPin-Priority: 900\n" | sudo tee /etc/apt/preferences.d/99nginx
```

Update apt and install nginx

```
sudo apt update
sudo apt install nginx
```

## Install Keycloak

> Work on **hserv**

### Prerequisites

Install jdk

```
sudo apt install default-jdk
```

Remove anacron

```
sudo apt remove anacron
```

Reboot the **hserv** machine

> ### Keycloak installation

Reviewer 2: ```suggestion
### Keycloak Installation
```

Go in base installation directory and get keycloak installation files (verify what is the last release)

```
cd ~/H/
wget https://github.com/keycloak/keycloak/releases/download/20.0.1/keycloak-20.0.1.zip
```

Extract the files

```
unzip keycloak-20.0.1.zip
```

Go to the bin directory and start keycloak in standalone mode

```
cd ~/H/keycloak-20.0.1/bin/
./kc.sh start-dev
```

Verify that Keycloak is accessible from **hserv** at the URL **"http://localhost:8080"**

Create the admin user as name **"admin"** and password **"1357Togo"**

![k6k01](https://github.com/MirtoBusico/assets-for-blogs/blob/main/k6k01.png)

Go to the administration console

![k6k02](https://github.com/MirtoBusico/assets-for-blogs/blob/main/k6k02.png)

Login and the “Master” realm appears. Note the Keycloak version

![k6k03](https://github.com/MirtoBusico/assets-for-blogs/blob/main/k6k03.png)

> ### Automatic Keycloak startup

Reviewer 2: ```suggestion
### Automatic Keycloak Startup
```

```
sudo mkdir /usr/share/nginx/k6k
sudo chmod 777 /usr/share/nginx/k6k
vi /usr/share/nginx/k6k/index.html
```

Put in index.html filke the "K6K" base document

```
<!DOCTYPE html>
<html>
<text>

<h1>K6K default page</h1>

</text>
</html>
```

In the directory **“/etc/nginx/conf.d”** create the file **“k6k.conf”**

```
cd /etc/nginx/conf.d
sudo vi k6k.conf
```

The file contains

```
server {

    listen 443 ssl http2;
    server_name k6k.h.net;
    root   /usr/share/nginx/k6k;

    access_log  /var/log/nginx/k6k.access.log;
    error_log  /var/log/nginx/k6k.error.log;

    ssl_certificate     /home/sysop/H/hservcerts/k6k.crt;
    ssl_certificate_key /home/sysop/H/hservcerts/k6k.key;

    location / {
        index  index.html index.htm;
    }

    error_page  404              /404.html;

    # redirect server error pages to the static page /50x.html
    #
    error_page   500 502 503 504  /50x.html;
    location = /50x.html {
        root   /usr/share/nginx/html;
    }

}
```

Restart Nginx

```
sudo systemctl restart nginx
```

Try to access “https://k6k.h.net” from a browser: the k6k base document will be showed

> ### Add keycloak reverse proxy

Reviewer 2: ```suggestion
### Add Keycloak Reverse Proxy
```

    proxy_set_header Host $host;
    proxy_set_header X-Real-IP $remote_addr;
    proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
    proxy_set_header X-Forwarded-Host $host;
    proxy_set_header X-Forwarded-Proto $scheme;

    location / {
        proxy_pass http://127.0.0.1:8080;
    }

    error_page  404              /404.html;

    # redirect server error pages to the static page /50x.html
    #
    error_page   500 502 503 504  /50x.html;
    location = /50x.html {
        root   /usr/share/nginx/html;
    }

}
```

Restart Nginx

```
sudo systemctl restart nginx
```

Rebuild keycloak for production

```
cd ~/H/keycloak-20.0.1/bin/
./kc.sh --verbose build
```

Change in **“/usr/lib/systemd/system”** the file named **“keycloak.service”** with this content

```
[Unit]
Description=keycloak service
After=network.service

[Service]
ExecStart=/home/sysop/H/keycloak-20.0.1/bin/kc.sh start --proxy edge --hostname-strict=false >/var/log/keycloak.log 2>&1
PIDFile=/var/run/keycloak.pid

[Install]
WantedBy=multi-user.target
```

Enable and activate the service

```
sudo systemctl daemon-reload
sudo systemctl restart keycloak
```

> ## Apisix api gateway

Reviewer 2: ```suggestion
## Apisix API Gateway
```

### Addresses for apisix-dashboard

The address used is the exsternal address o the **hsrv** machine

In the **"/etc/hosts"** file of any machine accessing the cluster through the nginx reverse proxy add the line

```
192.168.100.20 apisix apisix.h.net
```

In the DNS server on **hserv** add the apisix entry in the **“h.net”** DNS zone with address **“192.168.100.20”**

![dns02](https://github.com/MirtoBusico/assets-for-blogs/blob/main/dns02.png)

> ### Apisix deployment

Reviewer 2: ```suggestion
### APISIX Deployment
```
